### ISSUE TITLE
Reversed Timeline Doesn't Work with "immediateRender: false"

### ISSUE THREAD

**Issue by jpotterm** [greensock/GreenSock-JS#111]
If you create a timeline to play later, you'll need to use `immediateRender: false` so that it doesn't render the `from` tweens immediately:

```javascript
var t = new TimelineLite({paused: true});
t.fromTo('.one', 1, {opacity: 0}, {opacity: 1, immediateRender: false});
```

This works fine when you play it normally:

```
t.play()
```

But if you play it in reverse

```
t.progress(1).reverse();
```

then it will finish the animation by rendering the start frame again. A reversed timeline should normally finish looking like `progress(0)`, but instead it finishes looking like `progress(1)`.

Here's a [demo](http://jsfiddle.net/g4xr1031/).

**Comment by jackdoyle** [comment]
This is actually precisely how it is supposed to work. Let me explain...

Let's say obj.x starts at 0. 

We build a timeline that has a fromTo() that makes obj.x go from 100 to 200. Okay, so we'd expect obj.x to jump instantly from 0 to 100, then animate to 200. Simple. 

Now if we play all that in reverse, where should obj.x end? Remember, it started at 0, jumped to 100 and animated to 200, so when the timeline is done reversing it should arrive back at 0. After all, the timeline is the thing that made it jump initially so when it's "hands off" at the end, it should safely get it back to that spot, fully reversing what it did. 

Originally, we didn't have that behavior and we had a few people point out that it was incorrect and sometimes very frustrating. This was actually a very tricky thing to implement/correct. 

So again, it is working exactly as expected but if there's a particular behavior you're after, I'd be happy to help you get it using the existing API/behavior. There are a bunch of ways you could tackle it. If you must have immediateRender:false, maybe use a callback that does a set() at the beginning of the timeline (thus it's not nested and doesn't get reverted). If you need any more tips, definitely post in our forums at http://greensock.com/forums/ - it's pretty active over there :)

**Comment by jpotterm** [comment]
Thanks for the clear explanation and the idea for a workaround! I should be able to work around it just fine now that I know what's going on. It sounds like this has already been discussed at length and probably won't change, but I'll describe why I found this confusing in case it's helpful.

When the initial state automatically becomes part of the timeline it feels like magic because now the timeline has things that I didn't put there. Worse, this magically added piece of the timeline only runs sometimes: on the first `play` and on every `reverse`.

These are some other reasons this approach seems unpredictable for me as the user:

- What if I run the same timeline multiple times and the initial state is different each time? Why does the initial state only become part of the timeline on the first run?
- I would expect the following two lines to do the same thing. `immediateRender` shouldn't have any effect since they are both played right away. However, if the CSS doesn't have them starting at `x: 0` then one will jump at the end and the other won't.

```javascript
new TimelineLite().fromTo('.one', 1, {x: 0}, {x: 200}).progress(1).reverse();
new TimelineLite().fromTo('.one', 1, {x: 0}, {x: 200, immediateRender: false}).progress(1).reverse();
```

Again, I appreciate your explanation; it's very helpful to know why GSAP is doing this. Now that I know this isn't a bug, I'll take my questions to the forums.

Thanks!

**Comment by jackdoyle** [comment]
Hm, you said the behavior only occurs on the first play() and every reverse(). I'm struggling to see that. From what I can tell, it's very consistent. Perhaps you could show me what you mean in a simplified codepen? 

The reason immediateRender:true acts differently is because that's intentionally supposed to be a "set it right away and then forget it" feature that's extremely useful for intro animations. As if the setting was done outside of the timeline. Again, this is all very intentional and took engineering to implement. Perhaps you could describe how you want it to work and then I can shed some light on the down sides to that approach (if I remember correctly, we thought through a lot of this stuff and landed on this approach because it was the most intuitive overall for the greatest number of scenarios but any path we took had some sort of scenario where someone might see it as unintuitive).  

I'm open to considering a change - maybe you've got a solution that we hadn't thought of and it'd be a great fit.

**Comment by jpotterm** [comment]
Here's a [demo](http://jsfiddle.net/7fm7whua/1/) that illustrates what seems different about playing a timeline forward and backward. The blue and red squares have the same timelines. The only difference is that I'm playing one forward and one backwards. Every time you click it replays again. Both blocks start at `x: 100` via the CSS. When `t1.play(0)` is called the first time, the blue block jumps from 100 to 0 and then animates to 200, which is expected. However ever subsequent call to `t1.play(0)` just jumps to 0 and animates to 200 never jumping to 100. In contrast, the red block that's animated in reverse jumps to 100 after every `t2.reverse(1)` is finished animating.

The behavior I expect in both of these cases is shown in this [demo](http://jsfiddle.net/51ey7e4y/1/) (using native web animation in Chrome).

I can definitely see the usefulness of having `{immediateRender: true}` jump the timeline to the initial state before animating. What seems weird to me is that with `{immediateRender: false}` the jump somehow becomes part of the timeline so that reverse performs the jump every time it finishes. Another way of saying it is that I'd expect to get the same animation if I take a `fromTo` tween, swap the `fromVars` with the `toVars`, and play it in reverse. But this [demo](http://jsfiddle.net/ft3273ny/3/) shows that's not the case with `{immediateRender: false}`.

I don't see a use case for having reverse with `{immediateRender: false}` jump to the initial state, but that doesn't mean there's not one.

Thanks for being open to discussion, and I hope my explanation is clear enough!

**Comment by jackdoyle** [comment]
Yes, I totally see what you mean, but I'm not sure I'm explaining the counter-point well enough. Hm. Let me try again (not sure I'll do much better...)

If I implement the behavior you describe, it will cause things to look "wrong" in an even more common scenario. Construct a timeline that has a few nested timelines with some fromTo() calls in them, sequenced. Play it forward. It'll act one way. Now restart() and it'll play differently. 

Like in your demo, if halfway through the main timeline, a nested timeline has a fromTo() that makes the box jump from 100 to 0 and then animated to 200. Great. Now rewind to the beginning of that master timeline and...uh oh...the box would be at 0. Technically according to your tweening code that's accurate (depending how you look at it), but it feels very wrong to have the animation play completely differently the second time (the box never went back to 100 at the start). 

I'm really struggling to find some magic solution that would intelligently ascertain which behavior you'd want in each situation. Some would argue that the behavior you described is most intuitive. Others have definitely told me it's not and that they were really surprised when their timeline and differently after they restarted. Actually, the behavior you're describing is a lot easier to implement in the engine. It took a *lot* of effort to get the current behavior :)

Plus there's already a way you can get the behavior you want - just don't set immediateRender at all (true is the default for fromTo() tweens). Or use that onStart trick I mentioned.

**Comment by mystrdat** [comment]
Thank you both for explanations on this issue, quite a headache as the default behavior is sort of unexpected. I've rather replaced all tweens with `fromTo` so the property state is clear.
